### PR TITLE
Added searchable encryption functionality to Encryptonize.

### PR DESCRIPTION
**Pull request by AmalieDue**
### Description

Added searchable encryption functionality to Encryptonize.

Fixes DEV-98.

### Type(s) of Change (Split the PR if you check many)

- [x] Added user feature
- [ ] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist

- [ ] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [x] I have added/updated end-to-end tests for feature changes.

### General Checklist

- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

**Comment by AmalieDue** [comment]
I had to handle IEK differently compared to the other keys, because each keyword/id pair has its own cryptor and tagger. However, I do not know if I did it correctly..

File: encryptonize.go [hunk 1]
```diff
@@ -77,8 +81,8 @@ func New(keys Keys) (Encryptonize, error) {
 	if err != nil {
 		return Encryptonize{}, err
 	}
-
-	return Encryptonize{&objectCryptor, &accessCryptor, &tokenCryptor, &userCryptor, &groupCryptor}, nil
+	indexKey := keys.IEK
+	return Encryptonize{&objectCryptor, &accessCryptor, &tokenCryptor, &userCryptor, &groupCryptor, indexKey}, nil
```
Comment (philip-cc) [inline encryptonize.go:85]:
```suggestion
	return Encryptonize{&objectCryptor, &accessCryptor, &tokenCryptor, &userCryptor, &groupCryptor, keys.IEK}, nil
```

File: encryptonize.go [hunk 2]
```diff
@@ -430,3 +434,27 @@ func (e *Encryptonize) GetGroupData(authorizer *SealedUser, group *SealedGroup)
 
 	return plainGroup.Data, nil
 }
+
+////////////////////////////////////////////////////////
+//                       Index                        //
+////////////////////////////////////////////////////////
+
```
Comment (philip-cc) [inline encryptonize.go:441]: We're missing a method for creating a new index. That would also be a good place to explain the feature a bit (in the doc string).

File: encryptonize_test.go [hunk 3]
```diff
@@ -1320,3 +1321,49 @@ func TestSharingObjectPart4(t *testing.T) {
 		}
 	}
 }
+
+func TestAddToIndex(t *testing.T) {
+	enc := newTestEncryptonize(t)
+
+	index := NewIndex()
```
Comment (philip-cc) [inline encryptonize_test.go:1328]: `NewIndex` should not be exported, it should be exposed through the `Encryptonize` struct. Come to think of it, `SealedID` should also be private.

File: encryptonize.go (hunk 2 above)
Comment (AmalieDue) [reply encryptonize.go:440]: Fixed

File: encryptonize_test.go (hunk 3 above)
Comment (AmalieDue) [reply encryptonize_test.go]: Fixed

File: encryptonize.go [hunk 4]
```diff
@@ -430,3 +433,35 @@ func (e *Encryptonize) GetGroupData(authorizer *SealedUser, group *SealedGroup)
...
+// NewIndex creates an instance of index which is a building block of searchable encryption.
+// An index contains a mapping from labels to sealedIDs. A label is based on a keyword, and
+// a sealedID is based on the occurrence of that keyword in a specific ID (e.g. a document).
+// Hence, index makes it possible to manage which keywords are contained in which IDs.
```
Comment (philip-cc) [inline encryptonize.go:444]: I think this description is too focused on the implementation details, which are not really relevant to the caller. What the caller needs to know is that this creates a new encrypted index and that this index can be used to map "keywords" to "IDs" without revealing the mapping.

File: encryptonize.go [hunk 5]
```diff
@@ -430,3 +433,35 @@ func (e *Encryptonize) GetGroupData(authorizer *SealedUser, group *SealedGroup)
...
+// Add adds the keyword/id pair to Index i.
+func (e *Encryptonize) Add(keyword, id string, i *index) error {
+	if err := i.add(e.IndexKey, keyword, id); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Given a keyword, Search returns all decrypted ids (contained in Index i)
```
Comment (philip-cc) [inline encryptonize.go:458]: Docstrings should start with the name of the thing they document, e.g. "Search finds all IDs that contain the given keyword..."

File: encryptonize.go (hunk 4 above)
Comment (AmalieDue) [reply encryptonize.go]: Fixed

**Review by philip-cc — APPROVED** [review]
Ship it!

**Pull request merged**